File: leetcode238/leetcode238.cpp
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> pre(n, 1);
        vector<int> post(n, 1);
        for (int i = 1; i < nums.size(); i++) {
            pre[i] = nums[i] * pre[i - 1];
        }
        for (int i = n - 1; i >= 0; i--){
            post[i] = nums[i] * post[i + 1];
        }
        for (int i = 0; i < n; i++) {
            nums[i] = pre[i] * post[i];
        }
        return nums;
    }
};

File: leetcode238/leetcode238.py
class Solution:
    def productExceptSelf(self, nums: list[int]) -> list[int]:
        n = len(nums)
        pre = [1] * n
        post = [1] * n
        for i in range(1, n):
            pre[i] = nums[i - 1] * pre[i - 1]
        for i in range(n - 2, -1, -1):
            post[i] = nums[i + 1] * post[i + 1]
        for i in range(n):
            nums[i] = pre[i] * post[i]
        return nums
